### PR TITLE
Swipe store refactor

### PR DESCRIPTION
**Pull request by matejcik**
saves 14 kB of flash, which should hopefully fit us with debuglink again. 

summary of changes:
* each screen of a swipe flow now implements an object-safe trait `FlowComponentDynTrait`
* ...allowing SwipeFlow to store a Vec of those things...
* so that we can completely remove SwipeStore and the recursive generic expansion
* additionally, state controllers are reworked somewhat to make them (a) more compact and (b) safer

speaking of, there was a (seemingly?) unused 3rd screen in `confirm_reset_recovery`, please verify that the removal is ok

### CODE REVIEW

File: core/embed/rust/src/ui/flow/swipe.rs [hunk 1]
```diff
@@ -1,29 +1,97 @@
...
+
+/// Shortcut type for the concrete renderer passed into `render()` method.
+type RendererImpl<'a, 'alloc, 'env> = ScopedRenderer<'alloc, 'env, ConcreteRenderer<'a, 'alloc>>;
+
+/// Fully object-safe component-like trait for flow components.
+///
+/// This trait has no generic parameters:
+/// * it is instantiated with a concrete Renderer type, and
+/// * it requires the `FlowComponentTrait` trait to be implemented for _any_
+///   lifetimes.
```
Comment (mmilata) [inline core/embed/rust/src/ui/flow/swipe.rs:71]: Trying to understand Higher Rank Trait Bounds ... here the `for` is used to erase lifetimes from a trait. The cost of this is that any function argument with such lifetime has the shortest possible lifetime, i.e. becomes invalid immediately after the function returns. Correct?

**Review by mmilata — APPROVED** [review]
Neat, thanks!

File: core/embed/rust/src/lib.rs [hunk 2]
```diff
@@ -8,6 +8,7 @@
 #![allow(dead_code)]
 #![feature(lang_items)]
 #![feature(optimize_attribute)]
+//#![feature(trait_upcasting)]
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/embed/rust/src/lib.rs:11]: ah, that should have been deleted.

`trait_upcasting` lets you, for `trait X: Y`, cast from `dyn X` to `dyn Y`. A previous version of the code used it for the `Swipeable` trait, but the requirement went away at some point.

File: core/embed/rust/src/ui/flow/base.rs [hunk 3]
```diff
@@ -44,24 +43,71 @@ impl<Q> Decision<Q> {
...
+    /// this trait must remain object-safe and so can't refer to Self.
+    fn handle_event(&'static self, msg: FlowMsg) -> StateChange;
+
+    /// Page index of the current state.
+    fn index(&'static self) -> usize;
 }
+
+/// Helper trait for writing nicer flow logic.
+pub trait DecisionBuilder: FlowState + Sized {
+    #[inline]
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/embed/rust/src/ui/flow/base.rs:77]: i certainly hope they wouldn't :)
i put it in more as a semantic information that those are just shortcuts and there should not be anything non-inlineable here
but i can remove them if you like..?

File: core/embed/rust/src/ui/flow/swipe.rs (hunk 1 above)
Comment (matejcik) [reply core/embed/rust/src/ui/flow/swipe.rs:71]: That's a weird way to put it but essentially yes?
But note that this is the _default behavior_.
If you write a function accepting `impl Renderer`, without any lifetime parameters, the function also cannot rely on the argument existing beyond the scope of the function.

The issue here is that `FlowComponentTrait` has a lifetime parameter.
`Component` is fine with `render<'s>(&'s self, target: &mut impl Renderer<'s>)`, i.e., "render me with anything that lives as long as me"
But that's not object-safe. To be object-safe, we need to bind to a specific implementation of `Renderer` with a specific lifetime. That's the lifetime `'s` of `FlowComponentTrait`

buuuuut also, we `impl<'s> FlowComponentTrait<'s> for C: Component, R: Renderer<'s>`
so in other words, for any pre-specified `'s` and `R`, any `Component` implements `FlowComponentTrait` (which is obviously fine with Component, seeing as Component doesn't care about `'s`)

and the next step is `FlowComponentDynTrait`, saying "every implementer that implements `FlowComponentTrait<'s>` for any `'s` also implements `FlowComponentDynTrait`"
essentially circling back to what `Component` already does with `render<'s>(&'s self...)`

File: core/embed/rust/src/ui/flow/swipe.rs [hunk 4]
```diff
@@ -1,29 +1,97 @@
...
-/// Given a state enum and a corresponding FlowStore, create a Component that
-/// implements a swipe navigation between the states with animated transitions.
+use super::{base::StateChange, Swipable};
+
+use heapless::Vec;
+
+/// Component-like proto-object-safe trait.
+///
+/// This copies the Component interface, but it is parametrized by a concrete
+/// Renderer type, making it object-safe.
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/embed/rust/src/ui/flow/swipe.rs:30]: not here, but added two more comments 2748c642773d22291af152aabeea12b7da69e57c

File: core/embed/rust/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (matejcik) [reply core/embed/rust/src/lib.rs]: 6da327b114ec5ea65ce6b92235f59dd6b3e420d9

File: core/embed/rust/src/ui/flow/base.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmilata) [reply core/embed/rust/src/ui/flow/base.rs:77]: no need, was just wondering if it was an intention

File: core/embed/rust/src/ui/model_mercury/flow/confirm_reset_recover.rs [hunk 5]
```diff
@@ -104,23 +97,9 @@ impl ConfirmResetRecover {
...
-        let content_confirm = Frame::left_aligned(
-            TR::reset__title_create_wallet.into(),
-            SwipeContent::new(PromptScreen::new_hold_to_confirm()),
-        )
-        .with_footer(TR::instructions__hold_to_confirm.into(), None)
-        .with_swipe(SwipeDirection::Down, SwipeSettings::default())
-        .map(|msg| match msg {
-            FrameMsg::Content(()) => Some(FlowMsg::Confirmed),
-            _ => Some(FlowMsg::Cancelled),
-        });
```
[the remark this answers is not in the document]
Comment (obrusvit) [reply core/embed/rust/src/ui/model_mercury/flow/confirm_reset_recover.rs:116]: @mmilata This was most probably an omission after copying `confirm_reset_create.rs` which is very similar to this flow but without "Hold to confirm" screen. Also `TR::reset__title_create_wallet` is used supporting this.

Btw. thinking of ways to merge `confirm_reset_create.rs` and `confirm_reset_recover.rs` together.
